src/input/datetime.rs

```diff
@@ -496,6 +496,11 @@ impl TzInfo {
         None
     }
 
+    fn fromutc<'p>(&self, py: Python<'p>, dt: &'p PyDateTime) -> PyResult<&'p PyDateTime> {
+        dt.call_method1("__add__", (self.utcoffset(py, py.None().as_ref(py))?,))?
+            .extract()
+    }
```

Comment:
Being unable to resist a little golfing here:

- No need to take `py` separately as you can get it from `dt.py()`
- I don't think there's value in casting the return value to datetime, that's just spending CPU cycles

```suggestion
    fn fromutc<'py>(&self, dt: &'py PyDateTime) -> PyResult<&'py PyAny> {
        let py = dt.py();
        dt.call_method1("__add__", (self.utcoffset(py, py.None().as_ref(py))?,))
    }
```

(`self.utcoffset()` could actually get similar treatment)

Reply:
Just that it's all the same `py` so no need to have a separate argument for it. It's a bit like the following would be redundant:

```rust
fn f(x: i32, (_, y): (i32, i32)) { ... }

f(x, (x, y))  // passed x twice, ignored in second case
```

Reply:
Requesting it with the function argument is better than `Python::with_gil` if you don't already have it for sure.